ceph/ceph_admin/__init__.py

```diff
@@ -0,0 +1,750 @@
+
+        while checks:
+            checks -= 1
+
+            out, _ = self.shell(
+                remote=self.installer,
+                args=["ceph", "orch", "ps", "-f", "json-pretty"],
+            )
+
+            out = json.loads(out)
+            daemons = dict(
+                (i["daemon_id"], i["status_desc"])
+                for i in out
+                if i.get("daemon_type") == daemon
+            )
+
+            count = 0
+            for _id in ids:
+                if _id in daemons:
+                    if daemons[_id] == "running":
+                        count += 1
+
+            logger.info("%s/%s %s daemon(s) up...." % (count, len(ids), daemon))
+            if count == len(ids):
+                return True
+
+            logger.info("re-checking... %s times left", checks)
+            sleep(interval)
+
+        # Checkout service events, in case of processes not-running
+        out, _ = self.shell(
+            remote=self.installer,
+            args=["ceph", "orch", "ls", "-f", "json-pretty"],
+        )
+
+        for i in json.loads(out):
+            if daemon in i.get("service_type") and i["status"].get("running") == 0:
+                logger.exception("Service status(es) : %s" % i)
+                logger.exception("Service event(s): %s" % i["events"])
+
+        return False
+
+    def shell(self, remote, args, **kwargs):
+        """
+        Ceph orchestrator shell interface to run ceph commands
+
+        Args:
+            remote: remote host
+            args: list arguments
+            kwargs: key-value arguments
+
+        Returns:
+            out: stdout
+            err: stderr
+        """
+        if not remote:
+            remote = self.installer
+
+        cmd = ["sudo cephadm -v", "shell -- "] + args
+        [cmd.extend([k, v]) for k, v in kwargs]
+
+        out, err = remote.exec_command(
+            cmd=" ".join(cmd),
+            timeout=self.TIMEOUT,
+        )
+
+        out = out.read().decode().strip()
+        err = err.read().decode().strip()
+        logger.debug(out)
+        return out, err
+
+    def set_tool_repo(self):
+        """
+        Enable Ceph tool repo on every cluster node
+        """
+        for node in self.cluster.get_nodes():
+            node.exec_command(
+                cmd="sudo yum-config-manager --add"
+                " {}compose/Tools/x86_64/os/".format(self.config.get("base_url"))
+            )
+
+    def install_cephadm(self, **kw):
+        """
+        Install cephadm on installer node
+
+        Args:
+          kw: key-value configuration
+
+        kw:
+          upgrade: boolean # to upgrade cephadm RPM package
+
+        """
+        logger.info("Installing cephadm")
+        self.installer.exec_command(
+            sudo=True,
+            cmd="yum install cephadm -y --nogpgcheck",
+            long_running=True,
+        )
+
+        if kw.get("upgrade"):
+            self.installer.exec_command(cmd="sudo yum update metadata")
```

Comment:
Opt to set `sudo` as an argument to `exec_command`

Reply:
Done